src/storage/src/hummock/iterator/delete_range_iterator.rs

```diff
@@ -203,49 +199,85 @@ impl ForwardMergeRangeIterator {
     }
 }
 
-impl DeleteRangeIterator for ForwardMergeRangeIterator {
-    fn start_user_key(&self) -> UserKey<&[u8]> {
-        self.heap.peek().unwrap().start_user_key()
+impl ForwardMergeRangeIterator {
+    pub(super) fn next_until(&mut self, target_user_key: &UserKey<&[u8]>) {
+        while self.is_valid() && self.next_user_key().le(target_user_key) {
+            self.next();
+        }
     }
+}
 
-    fn end_user_key(&self) -> UserKey<&[u8]> {
-        self.heap.peek().unwrap().end_user_key()
+impl DeleteRangeIterator for ForwardMergeRangeIterator {
+    fn next_user_key(&self) -> UserKey<&[u8]> {
+        self.heap.peek().unwrap().next_user_key()
     }
 
     fn current_epoch(&self) -> HummockEpoch {
-        self.heap.peek().unwrap().current_epoch()
+        self.current_epochs
+            .range(..=self.read_epoch)
+            .last()
+            .map_or(HummockEpoch::MIN, |epoch| *epoch)
     }
 
     fn next(&mut self) {
-        let mut node = self.heap.peek_mut().expect("no inner iter");
-        node.next();
-        if !node.is_valid() {
-            // Put back to `unused_iters`
-            let node = PeekMut::pop(node);
-            self.unused_iters.push(node);
-        } else {
-            // This will update the heap top.
-            drop(node);
+        self.tmp_buffer
+            .push(self.heap.pop().expect("no inner iter"));
+        while let Some(node) = self.heap.peek() && node.is_valid() && node.next_user_key() == self.tmp_buffer[0].next_user_key() {
+            self.tmp_buffer.push(self.heap.pop().unwrap());
+        }
+        for node in &self.tmp_buffer {
+            let epoch = node.current_epoch();
+            if epoch != HummockEpoch::MAX {
```

Comment:
> IIUC, we are using `HummockEpoch::MAX` to indicate a non-existent range tomestaone between previous `next_user_key()` and current `next_user_key()`, which is counter-intuitive and requires deep understanding in the implementation before figuring it out. Can we make it more explicit by making `current_epoch` return `Option<HummockEpoch>`?

* `Option` will bring extra cost since `HummockEpoch` is not a non-zero type.
* The default value of min delete range is `MAX`. Isn't it intuitive?

Reply:
> Also, it is better to explicitly document `current_epoch()` indicates the range tomestone epoch between the previous `next_user_key()` and current `next_user_key()` to aid developers to understand the codes.

Fixed. I also added some explanation for the meaning of `HummockEpoch::MAX`.